contrib/docs/tutorials/cloudfoundry-deployment.md

```diff
@@ -0,0 +1,298 @@
+
+Once the push finishes, your Backstage instance is now running on Cloud Foundry! Navigate to the newly configured application's route at `https://APP_NAME.APPS_DOMAIN` in your browser.
+
+## Advanced
+
+### Authentication
+
+[UAA (User Account and Authentication)](https://docs.cloudfoundry.org/concepts/architecture/uaa.html) is the OAuth2 provider for Cloud Foundry. To use Cloud Foundry's UAA as the Backstage authentication provider, create a custom OIDC Provider.
+
+The following instructions are based on the [Backstage documentation for creating a custom OIDC provider](https://backstage.io/docs/auth/oidc/).
+
+#### Prerequisite
+
+The following steps utilize the UAA CLI (`uaac`). Installation and getting started can be found [here](https://docs.cloudfoundry.org/uaa/uaa-user-management.html). You will need to use `uaac` to login to your Cloud Foundry's UAA with a user capable of creating clients before continuing.
+
+#### Steps
+
+First create an API reference. This allows for dependency injection of our OIDC Auth API and gives a reference to our new provider API.
+
+In `packages/app/src/apis.ts`:
+```
+export const uaaOIDCAuthApiRef: ApiRef<
+  OpenIdConnectApi & ProfileInfoApi & BackstageIdentityApi & SessionApi
+> = createApiRef({
+  id: 'oidc',
+});
+```
+
+Next, create an API factory to create an instance of our custom OIDC API Provider.
+
+In `packages/app/src/apis.ts`:
+```
+export const apis: AnyApiFactory[] = [
+  ... // previous apis
+  createApiFactory({
+	api: uaaOIDCAuthApiRef,
+	deps: {
+  	discoveryApi: discoveryApiRef,
+  	oauthRequestApi: oauthRequestApiRef,
+  	configApi: configApiRef,
+	},
+	factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
+  	OAuth2.create({
+    	configApi,
+    	discoveryApi,
+    	oauthRequestApi,
+    	provider: {
+      	id: 'uaa-auth-provider',
+      	title: 'UAA auth provider',
+      	icon: () => null,
+    	},
+    	environment: configApi.getOptionalString('auth.environment'),
+    	defaultScopes: ['openid','profile','email'],
+  	}),
+  }),
+  ...
+];
+```
+
+The provider logic is in the `backend` package. This allows you to create custom resolvers to check for existing users and issue tokens. As an administrator of your Backstage instance, you have multiple opportunities here. You can load users into Backstage and connect catalog users to UAA users via an email, for example. You can also allow users who have no Backstage access yet to login. Both will be covered below.
+
+Create a `packages/backend/src/oidcProvider.ts` with the following:
+
+```
+import { createBackendModule } from '@backstage/backend-plugin-api';
+import {
+  authProvidersExtensionPoint,
+  createOAuthProviderFactory,
+} from '@backstage/plugin-auth-node';
+import { oidcAuthenticator } from '@backstage/plugin-auth-backend-module-oidc-provider';
+import { stringifyEntityRef } from '@backstage/catalog-model';
+import { NotFoundError } from '@backstage/errors';
+
+const uaaAuthProviderModule = createBackendModule({
```

Comment:
Should this maybe be `export const uaaAuthProviderModule`? then remove the `export default uaaAuthProviderModule;`?

Maybe a silly NIT but its usually the pattern I notice 😅